uvtools/dspec.py

```diff
@@ -146,6 +147,250 @@ def delay_filter(data, wgts, bl_len, sdf, standoff=0., horizon=1., min_dly=0.0,
+    order. As well as calculating a best-fit model, this will also return a 
+    copy of the data with flagged regions filled in ('in-painted') with the 
+    smooth solution. 
+    
+    Optionally, you can also add an uncorrelated noise realization on top of 
+    the smooth model in the flagged region.
+    
+    Parameters
+    ----------
+    data : array_like, complex
+        Complex visibility array as a function of frequency, with shape 
+        (Ntimes, Nfreqs).
+    
+    flags : array_like, bool
+        Boolean flags with the same shape as data.
+    
+    sigma : float or array_like
+        Noise standard deviation, in the same units as the data. If float, 
+        assumed to be homogeneous in frequency. If array_like, must have 
+        the same shape as the data.
+        
+        Note that the choice of sigma will have some bearing on how sensitive 
+        the fits are to small-scale variations.
+    
+    nmax: int
+        Max. order of Fourier modes to fit. A model with complex Fourier modes 
+        between [-n, n] will be fitted to the data, where the Fourier basis 
+        functions are ~ exp(-i 2 pi n nu / (Delta nu).
+    
+    add_noise : bool, optional
+        Whether to add an unconstrained noise realization to the in-painted areas. 
+        This uses sigma to set the noise standard deviation. Default: False.
+    
+    cn_guess : array_like, optional
+        Initial guess for the series coefficients of the first row of the 
+        input data. If None, zeros will be used. Default: None.
+    
+    use_linear : bool, optional
+        Whether to use a fast linear least-squares solver to fit the Fourier 
+        coefficients, or a slower generalized least-squares solver. 
+        Default: True.
+    
+    Returns
+    -------
+    model : array_like
+        Best-fit model, composed of a sum of Fourier modes. Same shape as the 
+        data.
+        
+    model_coeffs : array_like
+        Coefficients of Fourier modes, ordered from modes [-n, +n].
+    
+    data_out : array_like
+        In-painted data.
+    """
+    # Construct and cache Fourier basis operator (for speed)
+    nu = np.arange(data.shape[1])
+    L = nu[-1] - nu[0]
+    n = np.arange(-nmax, nmax+1)
+    F = np.array([np.exp(-1.j * _n * nu / L) for _n in n])
+    
+    # Array to store in-painted data
+    inp_data = np.zeros(data.shape, dtype=np.complex)
+    cn_array = np.zeros((data.shape[0], n.size), dtype=np.complex)
+    mdl_array = np.zeros(data.shape, dtype=np.complex)
+    
+    # Loop over array
+    cn_out = None
+    for i in range(data.shape[0]):
+        bf_model, cn_out, data_out = delay_filter_leastsq_1d(
+                                                data[i], flags[i], sigma=sigma, 
+                                                nmax=nmax, add_noise=add_noise, 
+                                                use_linear=use_linear,
+                                                cn_guess=cn_out, operator=F)
```

Comment:
Ah, I see, the `operator` argument is being pre-computed to save on computation for the 2d version. In that case, we should explicitly say that in the docs, and recommend against the user providing their own arbitrary basis.

Reply:
Done.